### PR TITLE
Signature algorithms extension support

### PR DESCRIPTION
**Pull request by ecimpsmain**
In our project we meet an issue where we had to provide the supported signature algorithms by the client. Otherwise the DTLS communication was not working. I have updated the library to add possibility to provide the supported signature algorithms.

### CODE REVIEW

**Comment by boaks** [comment]
Thanks a lot for your contribution!

The Eclipse organization wants to ensure, that everybody can use your contribution under the Eclipse Licenses. Therefore a ECA is required. Please check our [contribution guide](https://github.com/eclipse/californium/blob/master/CONTRIBUTING.md).

I y you get any problems with that, don't hesitate to contact me.

**Comment by boaks** [comment]
In short:
Please read the information in [ECA](http://www.eclipse.org/legal/ECA.php)
If you agree, "SIGN" a ECA. Menu on the right sidebar of that page (below "About Us" ... "ECA"). Use the e-Mail address you used for the commits.

Ammend the commit to add a short description e.g. "Add support for signature algorithm extension" and last, sign the commit by the text:

> Signed-off-by: Your Name <your@e-mail>

Similar as in [PR Commit - Text in the blue box at the top](https://github.com/eclipse/californium/pull/1224/commits/cb34315cc1888eee0cd82ed310bdd85b3af1d7a9)

File: scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/SignatureAlgorithmsExtension.java [hunk 1]
```diff
@@ -0,0 +1,111 @@
+package org.eclipse.californium.scandium.dtls;
```
Comment (boaks) [inline scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/SignatureAlgorithmsExtension.java]: For new file we use a [license header](https://github.com/eclipse/californium/blob/master/license_header_template.txt) 

`(company) name` is for the company, if this contribution was developed on a job for that company, or your name, if you developed it in your free time.

**Comment by boaks** [comment]
This seems to be the client side only. Do you use a other implementation on the server side? Or do you plan to contribute the server part as well in an other PR?

**Comment by boaks** [comment]
On remark: we currently plan to release Californium 2.2.0 mainly as baseline for Leshan 1.0 (other Eclipse project). Therefore we want to postpone all changes, which are not required for Leshan 1.0 to the time after we release this 2.2.0. The current schedule for that is the 24. March.

File: scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/SignatureAlgorithmsExtension.java (hunk 1 above)
Comment (ecimpsmain) [reply scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/SignatureAlgorithmsExtension.java]: Hi. :) I have added the license header.

**Comment by ecimpsmain** [comment]
> This seems to be the client side only. Do you use a other implementation on the server side? Or do you plan to contribute the server part as well in an other PR?

Yes, this is only client side. We use other implementation on the server side. We do not plan currently to contribute to the server part. :)

File: scandium-core/src/main/java/org/eclipse/californium/scandium/config/DtlsConnectorConfig.java [hunk 2]
```diff
@@ -1452,6 +1467,21 @@ public Builder setSupportedCipherSuites(String... cipherSuites) {
...
+		 * <p>
+		 * The connector will use these signature algorithms (in exactly the same
+		 * order) during the DTLS handshake.
+		 * 
+		 * @param supportedSignatureAlgorithms the supported signature algorithms in the order of
+		 *            preference
+		 * @return this builder for command chaining
+		 */
+		public Builder setSupportedSignatureAlgorithms(List<SignatureAndHashAlgorithm> supportedSignatureAlgorithms) {
+			config.supportedSignatureAlgorithms = supportedSignatureAlgorithms;
```
Comment (boaks) [inline scandium-core/src/main/java/org/eclipse/californium/scandium/config/DtlsConnectorConfig.java]: For the other list in the configuration `ListUtils.init(List)` is called. That checks for double and wraps the list into a "unmodifiableList". Unfortunately that also requires to add `hashCode` and `equals` in `SignatureAndHashAlgorithm`.

File: scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/ClientHello.java [hunk 3]
```diff
@@ -84,6 +84,12 @@
 	 * the client's first preference first.
 	 */
 	private List<CipherSuite> supportedCipherSuites = new ArrayList<>();
+	
+	/**
+	 * This is a list of the signature algorithms and hash functions supported and preferred
+	 * by the client
+	 */
+	private List<SignatureAndHashAlgorithm> supportedSignatureAndHashAlgorithms = new ArrayList<>();
```
Comment (boaks) [inline scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/ClientHello.java]: This list seems to be unused.

File: scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/SignatureAlgorithmsExtension.java [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+		
+		// Constructor ////////////////////////////////////////////////////
+
+		/**
+		 * Creates an instance using the signature algorithms and the hash algorithms codes
+		 * 
+		 * @param signatureAndHashAlgorithms
+		 */
+		public SignatureAlgorithmsExtension(SignatureAndHashAlgorithm[] signatureAndHashAlgorithms) {
```
Comment (boaks) [inline scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/SignatureAlgorithmsExtension.java]: Even if some other extension-implementations still use a array (they require a cleanup :-) ), in this case using a List makes it easier.

File: scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/SignatureAlgorithmsExtension.java [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+				}
+			}
+			sb.append(StringUtil.lineSeparator());
+			return sb.toString();
+		}
+
+		public List<SignatureAndHashAlgorithm> getSupportedSignatureAndHashAlgorithmIds() {
+			return Collections.unmodifiableList(signatureAndHashAlgorithms);
+		}
+}
```
Comment (boaks) [inline scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/SignatureAlgorithmsExtension.java]: If you use the Eclipse IDE, the project contains a "eclipse-formatter-profile.xml", please apply that.
If you don't use the Eclipse IDE, I apply the formatter in a separate commit after merging.

**Comment by boaks** [comment]
Generally I found only minor issues.
If you're interested, please adapt your PR. Otherwise I would adapt it on my own after merging it.
I prefer a "unmodifiable list" (so it could just be copied and doesn't require a clone) and remove the "short temporary used array" from the API. As I mentioned above, that requires `hashCode` and `equals` in `SignatureAndHashAlgorithm` to be implemented as well.

If you interested, you may also add a test for interoperability with openssl in demo-apps/cf-secure/test. It should not be to complicated to do that.  
 
AFAIK, scandium supports only ECDHE_ECDSA and so "{sha1,ecdsa}". You don't use a other combination with success?

File: scandium-core/src/main/java/org/eclipse/californium/scandium/config/DtlsConnectorConfig.java (hunk 2 above)
Comment (ecimpsmain) [reply scandium-core/src/main/java/org/eclipse/californium/scandium/config/DtlsConnectorConfig.java]: I added the `ListUtils.init(List)` call in the `public DtlsConnectorConfig build()` method and I have implemented the `hashCode` and `equals` methods in the `SignatureAndHashAlgorithm` class.

File: scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/ClientHello.java (hunk 3 above)
Comment (ecimpsmain) [reply scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/ClientHello.java]: Removed.

File: scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/SignatureAlgorithmsExtension.java (hunk 4 above)
Comment (ecimpsmain) [reply scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/SignatureAlgorithmsExtension.java]: Yes, it makes it easier, I thought it was a convention to do. :) Refactored.

File: scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/SignatureAlgorithmsExtension.java (hunk 5 above)
Comment (ecimpsmain) [reply scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/SignatureAlgorithmsExtension.java]: I applied the formatter.

**Comment by boaks** [comment]
LGTM

**Pull request merged**